File: Extra/Decode_String/decode__string.py
from typing import Optional, List
import heapq
from collections import defaultdict
import string


class Solution:
    def decodeString(self, s: str) -> str:
        if "[" not in s:
            return s

        nstart = 0
        while not s[nstart].isdigit():
            nstart += 1

        left = nstart
        while not s[left].isdigit():
            left += 1

        # find the first bracket
        start = 0
        while s[start] != "[":
            start += 1

        # n is the number of times to to multiple the []
        n = int(s[left:start])

        # find the matching ']'

        ptr = start + 1
        count = 1
        while count > 0:
            if s[ptr] == "[":
                count += 1
            elif s[ptr] == "]":
                count -= 1
            ptr += 1

        front = s[0:nstart]
        tmp = s[start + 1 : ptr - 1]
        mid = self.decodeString(tmp)
        end = self.decodeString(s[ptr:])

        return front + n * mid + end


if __name__ == "__main__":
    sol = Solution()

    s = "100[leetcode]"
    print(sol.decodeString(s))

    s = "3[a]2[bc]"
    print(sol.decodeString(s))

    s = "3[a2[c]]"
    print(sol.decodeString(s))

    print("Running Solution...")

File: NeetCode/Stack/Decode_String/Decode_String.cpp
// Decode_String.cpp
//
// Compile locally with:
// clang++ Decode_String.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;
class Solution
{
public:
    string decodeString(string s)
    {
        if (s.find('[') == string::npos)
        {
            return s;
        }

        int start = 0;

        while (start < s.size() &&
               !isdigit(static_cast<unsigned char>(s[start])))
        {
            start++;
        }

        int end = start;

        while (end < s.size() &&
               isdigit(static_cast<unsigned char>(s[end])))
        {
            end++;
        }

        string front = s.substr(0, start);

        // Characters from start up to, but not including, end
        int num = stoi(s.substr(start, end - start));

        // end points at '['
        int ptr = end + 1;
        int bracketCount = 1;

        while (bracketCount > 0)
        {
            if (s[ptr] == '[')
            {
                bracketCount++;
            }
            else if (s[ptr] == ']')
            {
                bracketCount--;
            }

            ptr++;
        }

        // ptr is now one position after the matching ']'
        string mid = s.substr(end + 1, ptr - end - 2);
        string back = s.substr(ptr);

        mid = decodeString(mid);
        back = decodeString(back);

        string result = front;

        for (int i = 0; i < num; i++)
        {
            result += mid;
        }

        result += back;

        return result;
    }
};

int main()
{
    cout << "Running Decode_String.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
